src/main/java/org/prebid/server/cache/CoreCacheService.java

```diff
@@ -385,26 +398,34 @@ private CachedCreative createJsonPutObjectOpenrtb(CacheBid cacheBid,
             bidObjectNode.put(BID_WURL_ATTRIBUTE, eventUrl);
         }
 
+        final String resolvedCacheKey = resolveCacheKey(accountId);
+
         final BidPutObject payload = BidPutObject.builder()
                 .aid(eventsContext.getAuctionId())
                 .type("json")
+                .key(resolvedCacheKey)
                 .value(bidObjectNode)
                 .ttlseconds(cacheBid.getTtl())
                 .build();
 
         return CachedCreative.of(payload, creativeSizeFromAdm(bid.getAdm()));
     }
 
-    private CachedCreative createXmlPutObjectOpenrtb(CacheBid cacheBid, String requestId, String hbCacheId) {
+    private CachedCreative createXmlPutObjectOpenrtb(CacheBid cacheBid,
+                                                     String requestId,
+                                                     String hbCacheId,
+                                                     String accountId) {
         final BidInfo bidInfo = cacheBid.getBidInfo();
         final Bid bid = bidInfo.getBid();
         final String vastXml = bid.getAdm();
 
-        final String customCacheKey = resolveCustomCacheKey(hbCacheId, bidInfo.getCategory());
+        final String resolvedCacheKey = resolveCacheKey(accountId, hbCacheId);
+        final String customCacheKey = resolveCustomCacheKey(resolvedCacheKey, bidInfo.getCategory());
 
         final BidPutObject payload = BidPutObject.builder()
                 .aid(requestId)
                 .type("xml")
+                .key(customCacheKey != null ? customCacheKey : resolvedCacheKey)
```

Comment:
Please, move key decision making logic to resolveCacheKey function, so that it will truly resolve cache key.